### PR TITLE
Consistently handle settings with multiple values

### PR DESCRIPTION
**Pull request by r-richardson**
As mentioned in: https://progress.opensuse.org/issues/166154

Screenshot From 2024-10-15 23-51-35

### CODE REVIEW

File: lib/OpenQA/Schema/ResultSet/Jobs.pm [hunk 1]
```diff
@@ -112,6 +112,13 @@ sub create_from_settings {
     my ($self, $settings, $scheduled_product_id) = @_;
 
     my %settings = %$settings;
+    foreach my $key (keys %settings) {
+        my $value = $settings{$key};
+
+        if (ref $value eq 'ARRAY' || ref $value eq 'HASH') {
+            $settings{$key} = encode_json($value);
+        }
+    }
```
Comment (okurz) [inline lib/OpenQA/Schema/ResultSet/Jobs.pm:121]: please try if this works

```suggestion
    my %settings = map { ref $settings->{$_} eq 'ARRAY' || ref $settings->{$_} eq 'HASH' ? encode_json($settings->{$_}) : $settings->{$_} } %$settings;
```

or if you want to use a helper variable for the value you could try

```suggestion
    my %settings = map { my $v = $settings->{$_}; ref $v eq 'ARRAY' || ref $v eq 'HASH' ? encode_json($v) : $v } %$settings;
```

**Review by okurz — CHANGES_REQUESTED** [review]

File: lib/OpenQA/Schema/ResultSet/Jobs.pm [hunk 2]
```diff
@@ -112,6 +112,13 @@ sub create_from_settings {
     my ($self, $settings, $scheduled_product_id) = @_;
 
     my %settings = %$settings;
+    foreach my $key (keys %settings) {
+        my $value = $settings{$key};
+
+        if (ref $value eq 'ARRAY' || ref $value eq 'HASH') {
+            $settings{$key} = encode_json($value);
```
Comment (Martchus) [inline lib/OpenQA/Schema/ResultSet/Jobs.pm:119]: Does this work with non-ASCII characters like umlauts? If not you might need to encode this to UTF-8.

File: lib/OpenQA/Schema/ResultSet/Jobs.pm (hunk 1 above)
Comment (Martchus) [reply lib/OpenQA/Schema/ResultSet/Jobs.pm:121]: I would keep using the helper variable. However, I also think the for-loop you wrote initially is fine.

Comment (okurz) [reply lib/OpenQA/Schema/ResultSet/Jobs.pm:121]: we need to think less procedural programming :)

File: lib/OpenQA/Schema/ResultSet/Jobs.pm (hunk 2 above)
Comment (r-richardson) [reply lib/OpenQA/Schema/ResultSet/Jobs.pm:119]: you're right, they are not printed correctly, i'll do that :+1:

File: lib/OpenQA/Schema/ResultSet/Jobs.pm (hunk 1 above)
Comment (r-richardson) [reply lib/OpenQA/Schema/ResultSet/Jobs.pm:121]: getting a `400 Bad Request` for either one of the two suggestions, shouldnt i be modifying `$settings` directly ?

Comment (perlpunk) [reply lib/OpenQA/Schema/ResultSet/Jobs.pm:121]: yes, I would work on the `%settings` hash, not the `$settings` hashref parameter directly

**Review by okurz — CHANGES_REQUESTED** [review]
TDD plz

File: lib/OpenQA/Schema/ResultSet/Jobs.pm [hunk 3]
```diff
@@ -171,6 +171,12 @@ sub create_from_settings {
         }
     }
 
+    for my $key (keys %settings) {
+        my $value = $settings{$key};
+        $settings{$key} = encode_json($value) if (ref $value eq 'ARRAY' || ref $value eq 'HASH');
+    }
```
Comment (okurz) [inline lib/OpenQA/Schema/ResultSet/Jobs.pm:177]:
```suggestion
    %settings = map { my $v = $settings{$_}; ref $v eq 'ARRAY' || ref $v eq 'HASH' ? encode_json($v) : $v } %settings;
```

Comment (perlpunk) [reply lib/OpenQA/Schema/ResultSet/Jobs.pm:177]: That doesn't work.
You need to iterate over the **keys** of the settings hash, and also you need to return the key **and** the value inside of the map.
We tried your suggestion together and found it not working.
I vote for keeping it like it is.

Comment (okurz) [reply lib/OpenQA/Schema/ResultSet/Jobs.pm:177]: so what is a better way to write code then? Using a loop is fine if we need to abort at some point but here we want an altered copy of the complete hash.

Comment (okurz) [reply lib/OpenQA/Schema/ResultSet/Jobs.pm:177]: I prefer the for loop over your one-liner suggestion as that looks like not the intended use case for map

File: lib/OpenQA/Schema/ResultSet/Jobs.pm [hunk 4]
```diff
@@ -171,6 +171,12 @@ sub create_from_settings {
         }
     }
 
+    for my $key (keys %settings) {
+        my $value = $settings{$key};
+        $settings{$key} = encode_json($value) if (ref $value eq 'ARRAY' || ref $value eq 'HASH');
```
Comment (kalikiana) [inline lib/OpenQA/Schema/ResultSet/Jobs.pm:176]: Would you mind adding some test coverage? The fact that this doesn't break anything suggests this is untested 😶

**Review by okurz — APPROVED** [review]

File: lib/OpenQA/Schema/ResultSet/Jobs.pm [hunk 5]
```diff
@@ -6,6 +6,7 @@ package OpenQA::Schema::ResultSet::Jobs;
 use Mojo::Base 'DBIx::Class::ResultSet', -signatures;
 use DBIx::Class::Timestamps 'now';
 use Date::Format 'time2str';
+use Encode;
```
Comment (perlpunk) [inline lib/OpenQA/Schema/ResultSet/Jobs.pm:9]: I prefer explicitly imported functions:
```suggestion
use Encode qw(decode_utf8);
```

File: t/10-jobs.t [hunk 6]
```diff
@@ -14,7 +14,7 @@ use FindBin;
 use lib "$FindBin::Bin/lib", "$FindBin::Bin/../external/os-autoinst-common/lib";
 use Mojo::Base -signatures;
 use autodie ':all';
-use Encode;
+use Encode qw(decode_utf8);
```
Comment (okurz) [inline t/10-jobs.t:17]:
```suggestion
```

**Review by okurz — APPROVED** [review]

**Review by perlpunk — APPROVED** [review]

**Comment by kalikiana** [comment]
Note this is still affected by [poo#167482](https://progress.opensuse.org/issues/167482) - maybe it didn't have the update yet and a rebase is enough

**Comment by kalikiana** [comment]
@mergifyio rebase

**Pull request merged**